pkg/provisioning/provision.go

```diff
@@ -0,0 +1,240 @@
+	var err error
+	oldPl, err := s.pipelineService.Get(*s.ctx, id)
+	if err == nil {
+		update = true
+		destMap, srcMap, err = s.copyStatesAndDeletePipeline(oldPl, config)
+		if err != nil {
+			return err
+		}
+	} else {
+		update = false
+	}
+
+	newPl, err := s.createPipeline(id, config)
+	if err != nil {
+		return err
+	}
+	for k, cfg := range config.Connectors {
+		err := s.createConnector(id, k, cfg, destMap, srcMap)
+		if err != nil {
+			return err
+		}
+		_, err = s.pipelineService.AddConnector(*s.ctx, newPl, k)
+		if err != nil {
+			return err
+		}
+		for k1, cfg1 := range cfg.Processors {
+			err := s.createProcessor(id, processor.ParentTypeConnector, k1, cfg1)
+			if err != nil {
+				return err
+			}
+			_, err = s.connectorService.AddProcessor(*s.ctx, k, k1)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	for k, cfg := range config.Processors {
+		err := s.createProcessor(id, processor.ParentTypePipeline, k, cfg)
+		if err != nil {
+			return err
+		}
+		_, err = s.pipelineService.AddProcessor(*s.ctx, newPl, k)
+		if err != nil {
+			return err
+		}
+	}
+
+	// set pipeline status
+	status := pipeline.StatusUserStopped
+	if config.Status == StatusRunning {
+		err := s.pipelineService.Start(*s.ctx, s.connectorService, s.processorService, newPl)
+		if err != nil {
+			return err
+		}
+		status = pipeline.StatusRunning
+	}
+	newPl.Status = status
+
+	return nil
+}
+
+func (s *Service) copyStatesAndDeletePipeline(pl *pipeline.Instance, config PipelineConfig) (map[string]connector.DestinationState, map[string]connector.SourceState, error) {
+	destMap := make(map[string]connector.DestinationState, 1)
+	srcMap := make(map[string]connector.SourceState, 1)
+
+	for k := range config.Connectors {
+		destState, srcState, err := s.getConnectorState(k)
+		if err != nil {
+			return nil, nil, err
+		}
+		destMap[k] = destState
+		srcMap[k] = srcState
+	}
+
+	err := s.pipelineService.Delete(*s.ctx, pl)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return destMap, srcMap, nil
+}
+
+func (s *Service) createPipeline(id string, config PipelineConfig) (*pipeline.Instance, error) {
+	cfg := pipeline.Config{
+		Name:        config.Name,
+		Description: config.Description,
+	}
+
+	pl, err := s.pipelineService.Create(*s.ctx, id, cfg, pipeline.ProvisionTypeConfig)
+	if err != nil {
+		return nil, err
+	}
+	return pl, nil
+}
+
+func (s *Service) createConnector(pipelineID string, id string, config ConnectorConfig, destMap map[string]connector.DestinationState, srcMap map[string]connector.SourceState) error {
+	cfg := connector.Config{
+		Name:       config.Name,
+		Plugin:     config.Plugin,
+		PipelineID: pipelineID,
+		Settings:   config.Settings,
+	}
+
+	connType := connector.TypeSource
+	if config.Type == TypeDestination {
+		connType = connector.TypeDestination
+	}
+
+	conn, err := s.connectorService.Create(*s.ctx, id, connType, cfg, connector.ProvisionTypeConfig)
+	if err != nil {
+		return err
+	}
+
+	if update {
```

Comment:
Maybe you don't need the update flag at all. If there's something in the map, we copy it, otherwise, we leave it as it is (empty state).